### PR TITLE
Deflake tests in `staging/src/k8s.io/kube-aggregator/pkg/apiserver`

### LINKED ISSUE #115858: [Flaking test] pull-kubernetes-unit - TestDirty fails sometimes
### Which jobs are flaking?

pull-kubernetes-unit sometimes fails with data race.
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115585/pull-kubernetes-unit/1622986339706212352

```sh
~/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver > go test -c -race ./
~/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver > stress ./apiserver.test -test.run=TestDirty                                       
                                                                                                     
/tmp/go-stress-20230217T225843-2898065946                                                                                                                                                                  
I0217 22:58:43.599838  409272 handler_discovery.go:392] Starting ResourceDiscoveryManager                                                                                                                  
--- FAIL: TestDirty (0.00s)                                                                                                                                                                                
    handler_discovery_test.go:165: service handler never pinged                                      
FAIL                                                                                                                                                                                                       
                                                                                                                                                                                                           
                                                                                                                                                                                                           
ERROR: exit status 1                                                                                                                                                                                       
                                                                                                                                                                                                           
                                                                                                                                                                                                           
/tmp/go-stress-20230217T225843-1296271475                                                                                                                                                                  
I0217 22:58:44.015272  409511 handler_discovery.go:392] Starting ResourceDiscoveryManager                                                                                                                  
==================                                                                                                                                                                                         
WARNING: DATA RACE                                                                                                                                                                                         
Write at 0x00c0003064bf by goroutine 93:                                                                                                                                                                   
  k8s.io/kube-aggregator/pkg/apiserver.TestDirty.func1()                                                                                                                                                   
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:154 +0x55                                                                            
  net/http.HandlerFunc.ServeHTTP()                                                                                                                                                                         
      /usr/local/go/src/net/http/server.go:2109 +0x4d                                                                                                                                                      
  net/http.(*timeoutHandler).ServeHTTP.func1()                                                                                                                                                             
      /usr/local/go/src/net/http/server.go:3407 +0xe1                                                                                                                                                      
                                                                                                                                                                                                           
Previous read at 0x00c0003064bf by goroutine 86:                                                                                                                                                           
  k8s.io/kube-aggregator/pkg/apiserver.TestDirty()                                                                                                                                                         
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:164 +0x433                                                                           
  testing.tRunner()                                                                                                                                                                                        
      /usr/local/go/src/testing/testing.go:1446 +0x216                                                                                                                                                     
  testing.(*T).Run.func1()                                                                                                                                                                                 
      /usr/local/go/src/testing/testing.go:1493 +0x47                                                                                                                                                      
                                                                                                                                                                                                           
Goroutine 93 (running) created at:                                                                   
  net/http.(*timeoutHandler).ServeHTTP()                                                                                                                                                                   
      /usr/local/go/src/net/http/server.go:3401 +0x5e5                                                                                                                                                     
  k8s.io/kube-aggregator/pkg/apiserver.(*discoveryManager).fetchFreshDiscoveryForService()                                                                                                                 
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:218 +0xf31
  k8s.io/kube-aggregator/pkg/apiserver.(*discoveryManager).syncAPIService()
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:346 +0x156
  k8s.io/kube-aggregator/pkg/apiserver.(*discoveryManager).Run.func1.1()
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:413 +0x12e
  k8s.io/kube-aggregator/pkg/apiserver.(*discoveryManager).Run.func1()
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:418 +0x38

Goroutine 86 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.g
```

### Which tests are flaking?

TestDirty

### Since when has it been flaking?

I think the test was added 3 months ago with the flaking state.

### Testgrid link

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-unit

### Reason for failure (if possible)

There are two issues.

https://github.com/kubernetes/kubernetes/blob/3165fe19b34359ed65efb692cf231afc4d257176/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go#L133-L167

First, updating and reading a bool-type variable is not thread-safe.

and there is also another problem checking the state before completely syncing it.

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig api-machinery

### PR DESCRIPTION
**Pull request by gjkim42**
#### What type of PR is this?

/kind failing-test
/kind flake

#### What this PR does / why we need it:

This PR addresses two issues in `handler_discovery_test.go`.
1. Updating and reading the bool-type variable is not thread-safe. We can simply fix this one.
2. (This is quite tricky to address) It is hard to know that `discoveryManager` completes the sync on all items in the queue.

More details on issue 2 below:
`waitForEmptyQueue` cannot guarantee that `discoveryManager` completes the sync on all items in the queue. It just guarantee that all items have been started to sync.

This PR adds `waitForQueueComplete` and implements `completerWorkqueue` to check if the workqueue is complete.

#### Which issue(s) this PR fixes:

Fixes #115858 

```sh
$ stress ./apiserver.test                    
5s: 0 runs so far, 0 failures       
10s: 12 runs so far, 0 failures     
15s: 24 runs so far, 0 failures     
...
3m50s: 445 runs so far, 0 failures
3m55s: 456 runs so far, 0 failures
4m0s: 461 runs so far, 0 failures
```

#### Does this PR introduce a user-facing change?

```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
in a super quick look, I can observe that you have getters to the controller cache, maybe that can give you a way to confirm that all items were processed ?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
Comment (alexzielenski) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:163]: While this blocks until all items in the queue have synced, I don't think it is the right general solution. `waitForEmptyQueue` needs to be modified to wait for queue to stop processing, otherwise the other tests that use `waitForEmptyQueue` will still be affected

**Comment by gjkim42** [comment]
@aojea

> in a super quick look, I can observe that you have getters to the controller cache, maybe that can give you a way to confirm that all items were processed ?

I think the way (using cache) looks not clear for now.

Actually, I am struggling to solve this issue with more general way. (the current fix looks just a hack for me)

I feel that there are a few workarounds,
1. waiting for the desired state with timeout in tests
2. adding processingCount to the controller
3. adding IsEmpty(Len is 0 and is not processing anything) to queue.Interface

3 looks better approach for me... what do you think?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
Comment (gjkim42) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:163]: I think so too. I am trying to solve this with more general way. See https://github.com/kubernetes/kubernetes/pull/115859#issuecomment-1434881423

**Comment by gjkim42** [comment]
> 3\. adding IsEmpty(Len is 0 and is not processing anything) method to queue.Interface

applied 3.

One concern is that I am not sure it is ok to add a method to `queue.Interface`

**Comment by gjkim42** [comment]
> > 3. adding IsEmpty(Len is 0 and is not processing anything) method to queue.Interface
> 
> applied 3.
> 
> One concern is that I am not sure it is ok to add a method to `queue.Interface`

There are some other implementations for `queue.Interface`...

**Comment by gjkim42** [comment]
added completerWorkQueue to check if the workqueue is complete for testing purpose.

**Comment by gjkim42** [comment]
/priority important-longterm

@alexzielenski 

Could you PTAL?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
Comment (alexzielenski) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:476]: This feels weird to me, but I know neither of these ever are called by discovery manager, since [`syncAPIService`](https://github.com/kubernetes/kubernetes/blob/40c5b1b7c37804f5c8656dffadf7f086daa0a9ba/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go#L344) unconditionally returns nil. (Discovery Manager refreshes all services on minute interval unconditionally, so does not make use of rate limiter)

I'm wondering if this or a future PR should refactor it to not use the `workqueue.RateLimitingInterface` and instead use normal `workqueue.Interface`. I think we can keep that in mind for a separate PR, to retain this PR's nice property of only changing test files.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
Comment (alexzielenski) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:382]: I created this test which deterministically caught the flake before your change, and is now working after your change. It makes the sync handler take a long time to catch the queue being empty, but handler still syncing. I think maybe it would be good to add:

```go
// Shows that waitForQueueComplete also waits for syncing to
// complete by artificially making the sync handler take a long time
func TestWaitForSync(t *testing.T) {
	pinged := atomic.Bool{}
	service := discoveryendpoint.NewResourceManager()
	aggregatedResourceManager := discoveryendpoint.NewResourceManager()

	aggregatedManager := newDiscoveryManager(aggregatedResourceManager)

	aggregatedManager.AddAPIService(&apiregistrationv1.APIService{
		ObjectMeta: metav1.ObjectMeta{
			Name: "v1.stable.example.com",
		},
		Spec: apiregistrationv1.APIServiceSpec{
			Group:   "stable.example.com",
			Version: "v1",
			Service: &apiregistrationv1.ServiceReference{
				Name: "test-service",
			},
		},
	}, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(3 * time.Second)
		pinged.Store(true)
		service.ServeHTTP(w, r)
	}))
	testCtx, cancel := context.WithCancel(context.Background())
	defer cancel()

	go aggregatedManager.Run(testCtx.Done())
	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))

	// immediately check for ping, since Run() should block for local services
	if !pinged.Load() {
		t.Errorf("service handler never pinged")
	}
}
```

**Review by alexzielenski — APPROVED** [review]
Thanks @gjkim42! I really like this solution with the test-specific completerWorkqueue. I think it solves this problem fairly cleanly. 

One minor comment about the RateLimitingInterface that may not need to be addressed here.

/lgtm
/assign @liggitt 
/cc @liggitt 
I think this test-only change will also need to be back ported to 1.26. My change to patch a race #115302 introduced this flake.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
Comment (gjkim42) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:419]: added it

**Comment by aojea** [comment]
> I think the way (using cache) looks not clear for now.

then forget my comment, I most probably misunderstood it, seems Alex is taking on the review , he understand this better than me

**Comment by gjkim42** [comment]
updated the commit message and rebased the PR.

**Comment by gjkim42** [comment]
/priority important-soon

`pull-kubernetes-unit` can be flaky.

**Comment by gjkim42** [comment]
> @alexzielenski can you take a final look at the most recent updates?

added a test that you suggested and changed the commit message.
Could you PTAL?

**Comment by gjkim42** [comment]
@liggitt 
Could you take a look?

> I think this test-only change will also need to be back ported to 1.26. My change to patch a race #115302 introduced this flake.

And we think that this needs to be backported.

**Comment by aojea** [comment]
> in a super quick look, I can observe that you have getters to the controller cache, maybe that can give you a way to confirm that all items were processed ?

```diff
diff --git a/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go b/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
index 9bd48972404..c07993779d7 100644
--- a/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
+++ b/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
@@ -18,6 +18,7 @@ package apiserver
 
 import (
        "context"
+       "fmt"
        "net/http"
        "net/http/httptest"
        "strconv"
@@ -44,12 +45,14 @@ func newDiscoveryManager(rm discoveryendpoint.ResourceManager) *discoveryManager
        return NewDiscoveryManager(rm).(*discoveryManager)
 }
 
-// Returns true if the queue of services to sync empty this means everything has
-// been reconciled and placed into merged document
-func waitForEmptyQueue(stopCh <-chan struct{}, dm *discoveryManager) bool {
+// Returns true when it reachs the number of the cached services
+func waitForCacheNumber(stopCh <-chan struct{}, n int, dm *discoveryManager) bool {
        return cache.WaitForCacheSync(stopCh, func() bool {
                // Once items have successfully synced they are removed from queue.
-               return dm.dirtyAPIServiceQueue.Len() == 0
+               dm.resultsLock.Lock()
+               defer dm.resultsLock.Unlock()
+               fmt.Println("DEBUG", len(dm.cachedResults), n)
+               return len(dm.cachedResults) == n
        })
 }
 
@@ -63,7 +66,6 @@ func TestBasic(t *testing.T) {
        service2.SetGroups(apiGroup2.Items)
        aggregatedResourceManager := discoveryendpoint.NewResourceManager()
        aggregatedManager := newDiscoveryManager(aggregatedResourceManager)
-
        for _, g := range apiGroup1.Items {
                for _, v := range g.Versions {
                        aggregatedManager.AddAPIService(&apiregistrationv1.APIService{
@@ -103,7 +105,7 @@ func TestBasic(t *testing.T) {
 
        go aggregatedManager.Run(testCtx.Done())
 
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 2, aggregatedManager))
 
        response, _, parsed := fetchPath(aggregatedResourceManager, "")
        if response.StatusCode != 200 {
@@ -159,7 +161,9 @@ func TestDirty(t *testing.T) {
        defer cancel()
 
        go aggregatedManager.Run(testCtx.Done())
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 1, aggregatedManager))
+       // time.Sleep(1 * time.Second)
+       fmt.Println("POST", aggregatedManager.apiServices, aggregatedManager.cachedResults)
 
        // immediately check for ping, since Run() should block for local services
        if !pinged.Load() {
@@ -211,7 +215,7 @@ func TestRemoveAPIService(t *testing.T) {
                aggregatedManager.RemoveAPIService(s.Name)
        }
 
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 0, aggregatedManager))
 
        response, _, parsed := fetchPath(aggyService, "")
        if response.StatusCode != 200 {
@@ -293,7 +297,7 @@ func TestLegacyFallback(t *testing.T) {
        defer cancel()
 
        go aggregatedManager.Run(testCtx.Done())
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 1, aggregatedManager))
 
        // At this point external services have synced. Check if discovery document
        // includes the legacy resources
@@ -362,7 +366,7 @@ func TestNotModified(t *testing.T) {
 
        // Important to wait here to ensure we prime the cache with the initial list
        // of documents in order to exercise 304 Not Modified
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 1, aggregatedManager))
 
        // Now add all groups. We excluded one group before so that AllServicesSynced
        // could include it in this round. Now, if AllServicesSynced ever returns
@@ -373,7 +377,7 @@ func TestNotModified(t *testing.T) {
        }
 
        // This would wait the full timeout on 1.26.0.
-       require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+       require.True(t, waitForCacheNumber(testCtx.Done(), 1, aggregatedManager))
 }
 ```

I tried the cache approach and looks easier than the queue

```
stress ./apiserver.test -test.run TestDirty
5s: 999 runs so far, 0 failures
10s: 2071 runs so far, 0 failures
15s: 3137 runs so far, 0 failures
20s: 4173 runs so far, 0 failures
25s: 5219 runs so far, 0 failures
30s: 6275 runs so far, 0 failures
35s: 7326 runs so far, 0 failures
40s: 8383 runs so far, 0 failures
45s: 9470 runs so far, 0 failures
50s: 10549 runs so far, 0 failures
```

**Comment by alexzielenski** [comment]
@aojea while that is a good approach I think if we in the future added a test for updating a pre-existing entry (which I'm likely to add very soon) this might not be sufficient if I'm reading correctly?

**Comment by liggitt** [comment]
looks like there are verify errors to resolve

/hold

**Comment by gjkim42** [comment]
> looks like there are verify errors to resolve

That is because https://github.com/kubernetes/kubernetes/pull/115770 has been merged which introduces a new test to fix.

Now, I rebased this PR and fixed the newly created test as well.

```
$ stress ./apiserver.test
5s: 0 runs so far, 0 failures       
10s: 12 runs so far, 0 failures     
15s: 12 runs so far, 0 failures
...
1h0m0s: 4582 runs so far, 0 failures
```

@alexzielenski @liggitt 

Could you re-LGTM on this?

**Comment by aojea** [comment]
> @aojea while that is a good approach I think if we in the future added a test for updating a pre-existing entry (which I'm now likely to add very soon) your alternative not be sufficient if I'm reading correctly?

@alexzielenski it was just an alternative because I saw your comment[ in the workqueue complete](https://github.com/kubernetes/kubernetes/pull/115859#discussion_r1110177336), but I don't have strong opinion, the current option LGTM too

**Pull request merged**
